README.md
# Homebridge Harman Kardon AVR

[![npm](https://img.shields.io/npm/v/homebridge-harman-kardon-avr?style=flat-square)](https://www.npmjs.com/package/homebridge-harman-kardon-avr) [![npm bundle size](https://img.shields.io/bundlephobia/min/homebridge-harman-kardon-avr?style=flat-square)](https://github.com/marcelkordek/homebridge-harman-kardon-avr)
[![GitHub last commit](https://img.shields.io/github/last-commit/marcelkordek/homebridge-harman-kardon-avr?style=flat-square)](https://github.com/marcelkordek/homebridge-harman-kardon-avr)

This is a plugin for [Homebridge](https://github.com/nfarina/homebridge) to control your **Harman Kardon AVR.** 

This plugin supports following functions:

- **Power Switch** (on/off)
- **Volume control** (volume up/volume down/mute)
- **Inputs** like STB, Cable/Sat, Game, Radio, Spotify etc.
- **Remote control:** native iOS Remote control

## Installation instructions
After [Homebridge](https://github.com/nfarina/homebridge) has been installed:

```
$ sudo npm install -g homebridge-harman-kardon-avr
```

## Basic configuration

 ```
{
	"bridge": {
		...
	},

	"accessories": [{
		"accessory": "harman-kardon-avr",
		"name": "AVR",
		"manufacturer": "Harman Kardon",
		"model_name": "AVR 161",
		"ip": "xx.x.x.xx",
		"port": "10025",
        "inputs": ["STB","Cable Sat","Game"],
        "interval": 5
	}]
}

 ```
 See [Example Config](https://github.com/marcelkordek/homebridge-harman-kardon-avr/blob/master/config-sample.json) for more details.


## Options

| **Attributes** | **Required** | **Usage** |
|------------|----------|-------|
| accessory | **Yes** | **Must be** harman-kardon-avr.   |
| name | **Yes** | **Unique Name** for the TV (AVR) Accessory.   |
| manufacturer | **No** | Manufacturer (optional - Default: Harman Kardon)   |
| model_name | **No** | Model (optional - Default: AVR 161)   |
| ip | **Yes** | IP adress from your AVR |
| port | **Yes** | **Must be** 10025 |
| inputs | **Yes** | An Array of Inputs (STB,CABLE SAT,GAME etc.) |
| interval | **Yes** | Polling interval in seconds _(Default: 5s)_ , 0 = disabled|

## Valid Inputs
* STB
* Cable Sat
* Disc
* Radio
* TV
* USB
* Game
* Media Server
* Home Network
* AUX

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v2.2.5](https://github.com/marcelkordek/homebridge-harman-kardon-avr/compare/v2.2.4...v2.2.5) - 2021-02-28

### Merged

- Bump node-fetch from 2.6.0 to 2.6.1 [`#7`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/pull/7)

### Commits

- fix homebridge 1.3 warning [`c5e4f3d`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/c5e4f3ddfae76eeaaa0a2afac48a9f3557e774ba)

## v2.2.4 - 2021-02-28

### Merged

- Bump lodash from 4.17.15 to 4.17.19 [`#6`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/pull/6)
- Bump acorn from 7.1.0 to 7.1.1 [`#2`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/pull/2)

### Commits

- update dependencies + fix homebridge 1.3 warning [`957b42c`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/957b42c4433788df457459f3a2856ae7780f57d5)
- TESTING: TV Integration, rewrite [`b87941d`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/b87941d12a8b94c1f16cf235c37e94f4a46b3bb1)
- Delete package-lock.json [`bde4ce3`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/bde4ce34e6e2110a6b83ebad221bb61399f49fc9)
- update index [`65e726b`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/65e726b2443e1094f74fb69d284ef8673190a292)
- add .gitignore [`740358d`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/740358d8150072a0aa9aeb877a664e53e6bdfc92)
- update index [`b4e1fba`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/b4e1fba7cff405fb9559ab780921bbde02797428)
- first commit [`cb454fd`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/cb454fd11ce5a055dab98cca8fd244dac9a31c50)
- [ADD] auto-changelog [`c6130fb`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/c6130fb3a7776632642c5394a22d467eb376d76f)
- update audio service [`cbc471f`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/cbc471f545e9473e56c21ca4c28030b68262a8a9)
- update index [`2cc7ed7`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/2cc7ed72f197aaaada1f951e59f044aafc348664)
- update services [`8376088`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/83760888c4ff91686a9462634e06bc21e24b3040)
- update [`b3bdf63`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/b3bdf63b6f0591652baf4c84b11d776ed974e0cf)
- Update index.js [`56b240b`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/56b240bed4c1b73c213043de816e7574123b4b60)
- [changed] scripts [`fd9a32a`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/fd9a32afd534d11f92a1438dba274b3ce0ac8709)
- Create config.schema.json [`4b3377d`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/4b3377db4137d3bdd83d0c285b1fc28d8d0a6f08)
- add keys [`352c6d6`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/352c6d64416913b9b8544ee985aa586fd513ad1e)
- updated CHANGELOG.md [`cfa8074`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/cfa8074bad6e5f229d4f1292c774bb02b386dafd)
- [changed] package.json [`b12c135`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/b12c1354cd0fc469fc00e1b365a1508c9614bfe9)
- updated CHANGELOG.md [`e1f45d8`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/e1f45d87d766b7b13d74444100b2ea5c48f7ff0e)
- update [`0d3cd2f`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/0d3cd2f4424148b8e6b96c7df3078801008dcda3)
- update readme [`1e3af17`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/1e3af171f3beabcebcdbc773e98d03a599c7180c)
- update index [`5b013fe`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/5b013fe370e76d3560ef4385c07e1308c26b7463)
- update package [`5b2822a`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/5b2822adfd208f6e2189c09c77e7fdb619e02497)
- add disable interval | fix powerState update [`4cb1b61`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/4cb1b613fe58d76f0b4723a7ee740683efe3d9a6)
- update [`c9325f3`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/c9325f32cfb62050918dbc4db038c1efcad9e013)
- update readme [`899d9b5`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/899d9b525ce15000d55475f489ae7d521636f282)
- update [`30d2557`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/30d2557365c37626b75888054b993f7f324cef2b)
- fix index.js [`678d4c0`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/678d4c0a5f0e9632496b5c0060d1d53572848dfd)
- update index.js [`f1f4912`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/f1f491237b46acd48bc27fdf101fb41532751e22)
- remove scoped package [`ca838d0`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/ca838d0efe257a840b80f5df860510d114dc6dbb)
- update [`b42fd5e`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/b42fd5ebfb07eda7a4ffeb293f83f93a41ecf45e)
- update package [`30e7a86`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/30e7a862dbd118a6b374933d709d4baa09d4456b)
- fix config.scheme.json [`1bfaeec`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/1bfaeec0569c8009653a5c5d551c8dd944d8fa46)
- update package.json [`c1c31fc`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/c1c31fc894dac1e5cdf86edb743381eb499ba272)
- update index [`d840ff5`](https://github.com/marcelkordek/homebridge-harman-kardon-avr/commit/d840ff5743e5a20bad87afa375505638890532d2)

index_.js
var Service, Characteristic
var exec = require('child_process').exec
var net = require('net')
var inherits = require('util').inherits

module.exports = function (homebridge) {
  Service = homebridge.hap.Service
  Characteristic = homebridge.hap.Characteristic

  // fixInheritance(HarmanKardonAVRAccessory.Volume, Characteristic);
  fixInheritance(HarmanKardonAVRAccessory.Mute, Characteristic)
  fixInheritance(HarmanKardonAVRAccessory.AudioService, Service)

  homebridge.registerAccessory('homebridge-harman-kardon-avr', 'harman-kardon-avr', HarmanKardonAVRAccessory)
}

function fixInheritance (subclass, superclass) {
  var proto = subclass.prototype
  inherits(subclass, superclass)
  subclass.prototype.parent = superclass.prototype
  for (var mn in proto) {
    subclass.prototype[mn] = proto[mn]
  }
};

function buildRequest (cmd, para) {
  var text = ''
  var payload = '<?xml version="1.0" encoding="UTF-8"?> <harman> <avr> <common> <control> <name>' + cmd + '</name> <zone>Main Zone</zone> <para>' + para + '</para> </control> </common> </avr> </harman>'
  text += 'POST HK_APP HTTP/1.1\r\n'
  text += 'Host: :' + this.ip + '\r\n'
  text += 'User-Agent: Harman Kardon AVR Controller/1.0\r\n'
  text += 'Content-Length: ' + payload.length + '\r\n'
  text += '\r\n'
  text += payload
  // console.log(text);
  return text
};

function HarmanKardonAVRAccessory (log, config) {
  this.log = log
  this.name = config.name
  this.ip = config.ip
  this.port = config.port
  this.model_name = config.model_name || 'AVR 161'
  this.manufacturer = config.manufacturer || 'Harman Kardon'
};

// custom characteristics
HarmanKardonAVRAccessory.Volume = function () {
  Characteristic.call(this, 'Volume', '00001001-0000-1000-8000-135D67EC4377')
  this.setProps({
    format: Characteristic.Formats.UINT8,
    unit: Characteristic.Units.PERCENTAGE,
    maxValue: 100,
    minValue: 0,
    minStep: 1,
    perms: [Characteristic.Perms.READ, Characteristic.Perms.WRITE, Characteristic.Perms.NOTIFY]
  })
  this.value = this.getDefaultValue()
}

HarmanKardonAVRAccessory.Mute = function () {
  Characteristic.call(this, 'Outlet In Use', '00000026-0000-1000-8000-0026BB765291')
  this.setProps({
    format: Characteristic.Formats.BOOL,
    perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
  })
  this.value = this.getDefaultValue()
}

HarmanKardonAVRAccessory.AudioService = function (displayName, subtype) {
  Service.call(this, displayName, '48a7057e-cb08-407f-bf03-6317700b3085', subtype)
  // this.addCharacteristic(HarmanKardonAVRAccessory.Volume);
  this.addOptionalCharacteristic(HarmanKardonAVRAccessory.Mute)
}

HarmanKardonAVRAccessory.prototype = {

  setPowerState: function (powerOn, callback) {
    var that = this
    if (powerOn) {
      var client = new net.Socket()
      client.connect(this.port, this.ip, function () {
        client.write(buildRequest('power-on'))
        client.destroy()
      })
    } else {
      var client = new net.Socket()
      client.connect(this.port, this.ip, function () {
        client.write(buildRequest('power-off'))
        client.destroy()
      })
    }
    client.on('error', function (err) {
      that.log('Error setting Powerstate: ' + err.message)
    })

    callback()
  },

  getPowerState: function (callback) {
    var that = this
    var state = false

    exec('ping -c 2 -W 1 ' + this.ip + " | grep -i '2 received'", function (error, stdout, stderr) {
      state = !!stdout
      that.log('Current state: ' + (state ? 'On.' : 'Off.'))
      callback(null, state)
    })
  },

  setInput: function (input, callback) {
    var that = this
    this.log('Set Input:' + input)
    var client = new net.Socket()
    if (input) {
      this.log('Change input to Cable/Sat')
      client.connect(this.port, this.ip, function () {
        client.write(buildRequest('source-selection', 'Cable Sat'))
        client.write(buildRequest('source-selection', 'Cable Sat'))
        client.destroy()
      })
    } else {
      this.log('Change input to STB')
      client.connect(this.port, this.ip, function () {
        client.write(buildRequest('source-selection', 'STB'))
        client.write(buildRequest('source-selection', 'STB'))
        client.destroy()
      })
    }
    client.on('error', function (err) {
      that.log('Error change input: ' + err.message)
    })

    callback()
  },

  identify: function (callback) {
    this.log('Identify requested!')
    callback() // success
  },

  getServices: function () {
    var availableServices = []
    var informationService = new Service.AccessoryInformation()

    availableServices.push(informationService)

    informationService
      .setCharacteristic(Characteristic.Manufacturer, this.manufacturer)
      .setCharacteristic(Characteristic.Model, this.model_name)
    // .setCharacteristic(Characteristic.SerialNumber, this.model_name);

    var outletService = new Service.Outlet(this.name)
    availableServices.push(outletService)

    outletService
      .getCharacteristic(Characteristic.On)
      .on('set', this.setPowerState.bind(this))
      .on('get', this.getPowerState.bind(this))

    var switchService = new Service.Switch('Sat/STB')
    availableServices.push(switchService)

    switchService
      .getCharacteristic(Characteristic.On)
      .on('set', this.setInput.bind(this))

    /* var audioService = new HarmanKardonAVRAccessory.AudioService('Input');
    availableServices.push(audioService);
      audioService
      .getCharacteristic(HarmanKardonAVRAccessory.Mute)
      .on('set', this.setInput.bind(this)); */

    return availableServices
  }
}


index.js
'use strict'

var Service, Characteristic, UUIDGen
var net = require('net')
var pkg = require('./package.json')
var tcpp = require('tcp-ping')
var Poller = require('./lib/poller')

module.exports = function (homebridge) {
  Service = homebridge.hap.Service
  Characteristic = homebridge.hap.Characteristic
  UUIDGen = homebridge.hap.uuid

  homebridge.registerAccessory('homebridge-harman-kardon-avr', 'harman-kardon-avr', HarmanKardonAVRAccessory)
}

function HarmanKardonAVRAccessory (log, config, api) {
  process.on('warning', e => log.warn(e.stack))

  // Config
  this.log = log
  this.api = api
  this.name = config.name
  this.ip = config.ip
  this.port = config.port || 10025
  this.model_name = config.model_name || 'AVR 161'
  this.manufacturer = config.manufacturer || 'Harman Kardon'
  this.inputs = config.inputs
  this.interval = config.interval || 0

  // Vars
  var that = this
  var powerOn = false
  this.enabledServices = []
  this.uuid = UUIDGen.generate(this.name)

  // Socket Connection
  this.client = new net.Socket()
  this.client.setNoDelay(true)
  this.client.setKeepAlive(true, 10000)

  // Information Service
  this.informationService = new Service.AccessoryInformation()
  this.informationService
    .setCharacteristic(Characteristic.Manufacturer, this.manufacturer)
    .setCharacteristic(Characteristic.Model, this.model_name)
    .setCharacteristic(Characteristic.SerialNumber, this.uuid)
    .setCharacteristic(Characteristic.FirmwareRevision, pkg.version)

  this.enabledServices.push(this.informationService)

  // Television Service
  // Add the actual TV Service and listen for change events from iOS.
  // We can see the complete list of Services and Characteristics in `lib/gen/HomeKitTypes.js`
  this.televisionService = new Service.Television(this.name, this.uuid)
  this.televisionService.category = this.api.hap.Categories.AUDIO_RECEIVER; 
  
  this.televisionService.setCharacteristic(Characteristic.ConfiguredName, this.name)

  this.televisionService.setCharacteristic(Characteristic.SleepDiscoveryMode, Characteristic.SleepDiscoveryMode.ALWAYS_DISCOVERABLE)

  this.televisionService
    .getCharacteristic(Characteristic.Active)
    .on('set', function (newValue, callback, context) {
      if (context === 'update') {
        newValue = !!newValue
        if (newValue !== powerOn) {
          that.log('updating Power State => ' + newValue)
          powerOn = newValue
         //callback(null, newValue)
        }
        callback(null)
        return
      }

      that.log('set Power State => ' + newValue)
      var power = newValue ? 'power-on' : 'power-off'
      powerOn = !!newValue

      that.command(power)

      callback(null)
    })
    .on('get', function (callback) {
      that.log('get State => ' + powerOn)
      callback(null, powerOn)
    })

  this.televisionService
    .getCharacteristic(Characteristic.ActiveIdentifier)
    .on('set', function (newValue, callback) {
      that.log('set Active Identifier => setNewValue: ' + newValue)

      var input = that.inputs[newValue]

      that.command('source-selection', input)

      callback(null)
    })

  this.televisionService
    .getCharacteristic(Characteristic.RemoteKey)
    .on('set', function (newValue, callback) {
      that.log('set Remote Key => setNewValue: ' + newValue)
      // ok = 8
      // hoch = 4
      // links = 6
      // rechts = 7
      // runter = 5
      // play/pause = 11
      // zurück = 9
      // i = 15
      var key
      switch (newValue) {
        case 9:
          key = 'back'
          break
        case 15:
          key = 'options'
          break
        case 8:
          key = 'ok'
          break
        case 4:
          key = 'up'
          break
        case 5:
          key = 'down'
          break
        case 6:
          key = 'left'
          break
        case 7:
          key = 'right'
          break
        default:
      }
      if (key) {
        that.command(key)
      }

      callback(null)
    })

  this.televisionService
    .getCharacteristic(Characteristic.PowerModeSelection)
    .on('set', function (newValue, callback) {
      that.log('set PowerModeSelection => setNewValue: ' + newValue)
      callback(null)
    })

  this.enabledServices.push(this.televisionService)

  // Speaker Service
  this.speakerService = new Service.TelevisionSpeaker(this.name + '_Speaker_' + this.uuid)

  this.speakerService
    .setCharacteristic(Characteristic.Active, Characteristic.Active.ACTIVE)
    .setCharacteristic(Characteristic.VolumeControlType, Characteristic.VolumeControlType.ABSOLUTE)

  this.speakerService
    .getCharacteristic(Characteristic.VolumeSelector)
    .on('set', function (newValue, callback) {
      that.log('set VolumeSelector => setNewValue: ' + newValue)
      var volume = newValue ? 'volume-down' : 'volume-up'

      that.command(volume)

      callback(null, newValue)
    })

  this.speakerService
    .getCharacteristic(Characteristic.Mute)
    .on('get', function (callback) {
      that.log('get Mute')
      callback(null, false)
    })
    .on('set', function (newValue, callback) {
      that.log('set Mute => setNewValue: ' + newValue)

      that.command('mute-toggle')

      callback(null)
    })

  this.televisionService.addLinkedService(this.speakerService)
  this.enabledServices.push(this.speakerService)

  // Input Service
  if (this.inputs && this.inputs.length !== 0) {
    for (var index in this.inputs) {
      index = index++
      that.log(index + ' -> ' + this.inputs[index])

      this.Identifier = index
      this.Name = this.inputs[index]
      this.InputSourceType = 'HDMI'
      this.InputDeviceType = 'Playback'
      this.CurrentVisibilityState = 'Shown'

      this.input = new Service.InputSource(this.Name, this.Name, that.uuid)
      this.input
        .setCharacteristic(Characteristic.Identifier, this.Identifier)
        .setCharacteristic(Characteristic.ConfiguredName, this.Name)
        .setCharacteristic(
          Characteristic.IsConfigured,
          Characteristic.IsConfigured.CONFIGURED
        )
        .setCharacteristic(
          Characteristic.InputSourceType,
          Characteristic.InputSourceType[this.InputSourceType]
        )
        .setCharacteristic(
          Characteristic.InputDeviceType,
          Characteristic.InputDeviceType[this.InputDeviceType]
        )
        .setCharacteristic(
          Characteristic.CurrentVisibilityState,
          Characteristic.CurrentVisibilityState[this.CurrentVisibilityState]
        )
        // .setCharacteristic(Characteristic.TargetVisibilityState, Characteristic.TargetVisibilityState[this.TargetVisibilityState])
        .on('set', function (newValue, callback) {
          that.log('set InputSource => setNewValue: ' + newValue)

          var input = this.inputs[index]

          that.command('source-selection', input)
          callback(null);
        })
      this.televisionService.addLinkedService(this.input)
      this.enabledServices.push(this.input)
    }
  }

  // Poller
  const poller = new Poller(this.interval * 1000)

  // Wait till the timeout sent our event to the EventEmitter
  poller.onPoll(() => {
    // that.log('triggered');
    tcpp.probe(that.ip, 8080, function (err, available) {
      if (err) {
        that.log(err)
        return
      }
      // that.log('Available: ', available)
      that.televisionService
        .getCharacteristic(Characteristic.Active)
        .setValue(available, undefined, 'update')
      poller.poll() // Go for the next poll
    })
  })
  // Initial start
  if (this.interval !== 0) {
    this.log('Start polling')
    poller.poll()
  } else {
    this.log('Polling disabled')
  }
}

// https://github.com/KarimGeiger/HKAPI
HarmanKardonAVRAccessory.prototype.buildRequest = function (cmd, param) {
  var request = ''
  var payload = '<?xml version="1.0" encoding="UTF-8"?> <harman> <avr> <common> <control> <name>' + cmd + '</name> <zone>Main Zone</zone> <para>' + param + '</para> </control> </common> </avr> </harman>'
  request += 'POST HK_APP HTTP/1.1\r\n'
  request += 'Host: :' + this.ip + '\r\n'
  request += 'User-Agent: Harman Kardon AVR Controller/1.0\r\n'
  request += 'Content-Length: ' + payload.length + '\r\n'
  request += '\r\n'
  request += payload

  cmd = cmd | ''
  param = param | ''
  this.log('Build Request Command: ' + cmd + ' ' + param)
  return request
}

HarmanKardonAVRAccessory.prototype.command = function (cmd, param) {
  var that = this

  that.log('Socket Writable: ' + this.client.writable)

  if (this.client.writable) {
    this.client.write(that.buildRequest(cmd, param))
  } else {
    this.client.connect(this.port, this.ip, function () {
      that.client.write(that.buildRequest(cmd, param))
      // that.client.destroy()
    })
  }
  this.client.on('data', function (data) {
    that.log('Client Data: ' + data.toString())
  })

  this.client.on('error', function (err) {
    that.log('Error setting State: ' + err.message)
    that.client.end()
    that.client.destroy()
  })
}

HarmanKardonAVRAccessory.prototype.identify = function (callback) {
  this.log('Identify requested!')
  callback()
}

HarmanKardonAVRAccessory.prototype.getServices = function () {
  return this.enabledServices
}
